### PR TITLE
Atmospheric forcing

### PR DESCRIPTION
**Pull request by NoraLoose**
Completes step 4 in #1.

This adds three classes: `AtmosphericForcing`, `SWRCorrrection`, and `ForcingDataset`, where the first two are the ones the user will interact with. I added an example notebook to the documentation that illustrates how the classes can be used in practice.

### ToDo
There are two things left to do, and I will open two separate issues to address these:

- [ ] **Test performance** and **profile memory** on different machines and clusters, and make recommendations to the user for what kind of dask clusters to use: (1) thread-based vs. process based, (2) how many workers for a given total amount of memory (making sure every worker has enough memory)
- [ ] Change **interpolation scheme** for zonal and meridional wind, and the SWR correction to mimic the MATLAB results more closely, see below.

### Python vs. MATLAB-based results
Both the `roms-tools` and the MATLAB-based scripts create 7 atmospheric forcing fields that are necessary to run a ROMS simulation. To create these fields both follow the same two-step procedure:
1. Fill in NaNs over land by spreading values from the ocean into land through diffusion. 
2. Interpolate from ERA5 to ROMS grid

Here is a comparison for all 7 fields for a randomly chosen time step. For each atmospheric field, there are 4 panels. 
* The left lower panel shows the difference between python and matlab over land and ocean. There are significant differences over land due to the fact that python vs. matlab use different boundary conditions for their diffusion-based NaN filling.
* Since we only care about differences over the *ocean*, the right lower panel masks out land. This is the important panel to look at.

We make the following observation:
 * For `Tair`, `qair`, `rain`, `lwrad`, the `roms-tools` and the MATLAB-based scripts lead to very similar results over the ocean. The reason is that both solutions use linear interpolation.
 *  There are significant differences for `uwnd`, `vwnd`, and `swrad` because for the time being `roms-tools` uses linear interpolation while the MATLAB-based scripts use a [modified akima scheme](https://www.mathworks.com/help/matlab/ref/makima.html).

comparison_Tair
comparison_qair
comparison_rain
comparison_lwrad
comparison_swrad
comparison_uwnd
comparison_vwnd

### CODE REVIEW

**Comment by NoraLoose** [comment]
@TomNicholas pinging you in case you have any thoughts on this PR. But no rush if you are busy with other stuff.

**Review by TomNicholas — COMMENTED** [review]
Some minor comments for now

File: roms_tools/setup/atmospheric_forcing.py [hunk 1]
```diff
@@ -0,0 +1,881 @@
...
+from roms_tools.setup.grid import Grid
+from datetime import datetime, timedelta
+import glob
+import numpy as np
+from typing import Optional, Dict, Union
+from scipy.sparse import spdiags, coo_matrix
+from scipy.sparse.linalg import spsolve
+from roms_tools.setup.fill import lateral_fill
+import warnings
+import calendar
```
[the remark this answers is not in the document]
Comment (NoraLoose) [reply roms_tools/setup/atmospheric_forcing.py:13]: Would it be useful to use `pre-commit` to run `isort` and other linting tools? https://pycqa.github.io/isort/docs/configuration/pre-commit.html

File: roms_tools/setup/atmospheric_forcing.py [hunk 2]
```diff
@@ -0,0 +1,881 @@
...
+
+        # Check if the file exists
+        matching_files = glob.glob(self.filename)
+        if not matching_files:
+            raise FileNotFoundError(f"No files found matching the pattern '{self.filename}'.")
+
+        # Load the dataset
+        with dask.config.set(**{'array.slicing.split_large_chunks': False}):
+            # initially, we wawnt time chunk size of 1 to enable quick .nan_check() and .plot() methods for AtmosphericForcing 
+            ds = xr.open_mfdataset(self.filename, combine='nested', concat_dim=self.dim_names["time"], chunks={self.dim_names["time"]: 1})
```
[the remark this answers is not in the document]
Comment (NoraLoose) [reply roms_tools/setup/atmospheric_forcing.py]: Thanks! Done via commit [0556bf1](https://github.com/CWorthy-ocean/roms-tools/pull/41/commits/0556bf1aaaab0def4f50618e5b8528ea8f051df1).

File: roms_tools/setup/atmospheric_forcing.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (TomNicholas) [reply roms_tools/setup/atmospheric_forcing.py:13]: Definitely! `ruff` does a lot of the formatting as one tool now though, e.g. 

https://github.com/zarr-developers/VirtualiZarr/blob/main/.pre-commit-config.yaml

[the remark this answers is not in the document]
Comment (NoraLoose) [reply roms_tools/setup/atmospheric_forcing.py]: Ok, I have run some linting tools on the `main` branch (PR https://github.com/CWorthy-ocean/roms-tools/pull/46) as well as this branch (commit [5b558e3](https://github.com/CWorthy-ocean/roms-tools/pull/41/commits/5b558e3494ca83b0f29a04c0248fe468909ee236)). `Pre-commit` now also runs some linting tools as part of the test suite (see checks below). Thanks for this suggestion - this repo really needed a clean-up!

[the remark this answers is not in the document]
Comment (TomNicholas) [reply roms_tools/setup/atmospheric_forcing.py]: Excellent! @dafyddstephenson you might want to steal what Nora has done here for the `C-Star` repo.

**Comment by NoraLoose** [comment]
@TomNicholas do we want to wait until the dask deployment issue is fixed to merge this PR? Maybe we should defer this to another PR?

**Comment by TomNicholas** [comment]
> @TomNicholas do we want to wait until the dask deployment issue is fixed to merge this PR? Maybe we should defer this to another PR?

As long as this code runs successfully for some dask setup somewhere, we can separate out the problem of deploying dask on other machines. So if this runs on Casper/your laptop that's good enough to merge this IMO.

Generally merging something which works first, then creating separate issues to track getting it to run in other setups or improving performance is a good idea.

**Comment by NoraLoose** [comment]
Sounds good! In my latest test, it actually ran on Perlmutter as well (with the default `LocalCluster()`), but there are still some weird things happening if other dask deployment strategies are used. I will try to document those in a separate issue.

**Pull request merged**